File: client/src/mocaclient/Encoding/Xml/XmlRequestEncoder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using RedPrairie.MOCA.Client.Interfaces;

namespace RedPrairie.MOCA.Client.Encoding.Xml
{
    /// <summary>
    /// A request encoder that creates an XML document.
    /// </summary>
    public class XmlRequestEncoder
    {
        /// <summary>
        /// Encodes the request.
        /// </summary>
        /// <param name="command">The command.</param>
        /// <param name="sessionId">The session id.</param>
        /// <param name="envString">The environment string.</param>
        /// <param name="context">The context argument list.</param>
        /// <param name="args">The explicit argument list. This generally should not be set.</param>
        /// <param name="autoCommit">if set to <c>true</c> auto commit the transaction.</param>
        /// <param name="writer">The writer.</param>
        public static void EncodeRequest(string command, string sessionId, string envString, IEnumerable<IMocaArgument> context, IEnumerable<IMocaArgument> args, bool autoCommit, StringBuilder writer)
        {

            writer.AppendFormat("<moca-request autocommit=\"{0}\">", autoCommit);

            if (sessionId != null)
            {
                writer.Append("<session id=\"");
                writer.Append(sessionId);
                writer.Append("\"/>");
            }
            if (envString != null)
            {
                writer.Append("<environment>");

                writer.Append(envString);
                writer.Append("</environment>");
            }

            if (context != null)
            {
                writer.Append("<context>");
                WiteArguments(context, writer);
                writer.Append("</context>");
            }

            if (args != null)
            {
                writer.Append("<args>");
                WiteArguments(args, writer);
                writer.Append("</args>");
            }

            if (command != null)
            {
                writer.Append("<query>");
                EscapeXml(command, writer);
                writer.Append("</query>");
            }

            writer.Append("</moca-request>");
        }

        /// <summary>
        /// Builds the XML environment string.
        /// </summary>
        /// <param name="env">The environment variables to encode.</param>
        /// <returns>An encoded environment string.</returns>
        public static String BuildXmlEnvironmentString(IDictionary<string, string> env)
        {
            var buf = new StringBuilder();
            foreach (var entry in env)
            {
                var name = entry.Key;
                var value = entry.Value;

                try
                {
                    buf.Append("<var name=\"");
                    EscapeXml(name.ToUpperInvariant(), buf);
                    buf.Append("\" value=\"");
                    EscapeXml(value, buf);
                    buf.Append("\"/>");
                }
                catch (IOException)
                {
                    // Could be thrown by EscapeXml, but won't because we're writing
                    // to a StringBuilder.
                }
            }

            return buf.ToString();
        }

        /// <summary>
        /// Escapes the XML.
        /// </summary>
        /// <param name="s">The string to escape.</param>
        /// <param name="writer">The stream writer output.</param>
        private static void EscapeXml(String s, StringBuilder writer)
        {

            if (String.IsNullOrEmpty(s))
                return;

            for (var i = 0; i < s.Length; i++)
            {
                var c = s[i];
                switch (c)
                {
                    case '&':
                        writer.Append("&amp;");
                        break;
                    case '<':
                        writer.Append("&lt;");
                        break;
                    case '>':
                        writer.Append("&gt;");
                        break;
                    case '"':
                        writer.Append("&quot;");
                        break;
                    case '\'':
                        writer.Append("&apos;");
                        break;
                    case ' ':
                    case '\t':
                    case '\n':
                    case '\r':
                        writer.Append(c);
                        break;
                    default:
                        if (Char.IsControl(c))
                        {
                            // Any non-whitespace control characters can appear as themselves, as long
                            // as they are escaped as character entities.  The exception, in Unicode 1.0,
                            // is that C0 control characters (0x00 .. 0x1f) cannot appear in any form.
                            // In order to allow strings that contain C0 control characters, we must move
                            // the entire protocol to XML 1.1.
                            if (c >= 0x00 && c <= 0x1f)
                            {
                                // Output the special "not a character" character
                                writer.Append('\ufffd');
                            }
                            else
                            {
                                writer.AppendFormat("&#x{0:x};", (int)c);
                            }
                        }
                        else
                        {
                            writer.Append(c);
                        }
                        break;
                }
            }
        }

        private static void WiteArguments(IEnumerable<IMocaArgument> args, StringBuilder writer)
        {
            foreach (var arg in args)
            {
                writer.Append("<field name=\"");
                EscapeXml(arg.Name, writer);
                writer.AppendFormat("\" type=\"{0}\" oper=\"{1}\"", 
                            arg.Type,  GetOperator(arg.Operator));

                object value = arg.Value;
                if (value == null)
                {
                    writer.Append(" null=\"true\"/>");
                }
                else
                {
                    writer.Append(">");
                    MocaType type = MocaType.LookupClass(value.GetType());

                    if (type == MocaType.BOOLEAN ||
                        type == MocaType.STRING ||
                        type == MocaType.INTEGER ||
                        type == MocaType.DOUBLE)
                    {
                        EscapeXml(arg.Value.ToString(), writer);  
                    }
                    else if (type == MocaType.DATETIME)
                    {
                        writer.Append(((DateTime) value).ToString(EncodingBase.DATE_TIME_FORMAT));
                    }
                    else if (type == MocaType.RESULTS ||
                             type == MocaType.BINARY)
                    {
                        //TODO: SUPPORT Results and Binary
                    }

                    writer.Append("</field>");
                }
            }
        }

        /// <summary>
        /// Gets the operator.
        /// </summary>
        /// <param name="mocaOperator">The moca operator.</param>
        /// <returns></returns>
        private static string GetOperator(MocaOperator mocaOperator)
        {
            switch (mocaOperator)
            {
                case MocaOperator.Equal:
                    return "EQ";
                case MocaOperator.NotEqual:
                    return "NE";
                case MocaOperator.GreaterThan:
                    return "GT";
                case MocaOperator.GreaterThanEqual:
                    return "GE";
                case MocaOperator.LessThan:
                    return "LT";
                case MocaOperator.LessThanEqual:
                    return "LE";
                case MocaOperator.Like:
                    return "LIKE";
                case MocaOperator.NotLike:
                    return "NOTLIKE";
                case MocaOperator.RawClause:
                    return "RAWCLAUSE";
                case MocaOperator.NamedClause:
                    return "NAMEDCLAUSE";
                default:
                    return "EQ";
            }
        }
    }
}

File: src/main/java/core/com/redprairie/moca/client/XMLRequestEncoder.java
/*
 *  $URL$
 *  $Revision$
 *  $Author$
 *  $Date$
 *  
 *  $Copyright-Start$
 *
 *  Copyright (c) 2016
 *  Sam Corporation
 *  All Rights Reserved
 *
 *  This software is furnished under a corporate license for use on a
 *  single computer system and can be copied (with inclusion of the
 *  above copyright) only for use on such a system.
 *
 *  The information in this document is subject to change without notice
 *  and should not be construed as a commitment by Sam Corporation.
 *
 *  Sam Corporation assumes no responsibility for the use of the
 *  software described in this document on equipment which has not been
 *  supplied or approved by Sam Corporation.
 *
 *  $Copyright-End$
 */

package com.redprairie.moca.client;

import java.io.IOException;
import java.util.Date;
import java.util.Map;

import com.redprairie.moca.MocaArgument;
import com.redprairie.moca.MocaOperator;
import com.redprairie.moca.MocaResults;
import com.redprairie.moca.MocaType;
import com.redprairie.moca.util.DateUtils;

/**
 * TODO Class Description
 * 
 * <b><pre>
 * Copyright (c) 2016 Sam Corporation
 * All Rights Reserved
 * </pre></b>
 * 
 * @author dinksett
 * @version $Revision$
 */
public class XMLRequestEncoder {
    
    /**
     * @param command
     * @param context
     * @param args
     * @param out
     * @throws IOException
     */
    public static void encodeRequest(String command, String sessionId, String envString, 
                              boolean autoCommit, boolean remote, 
                              MocaArgument[] context, MocaArgument[] args,
                              Appendable out) throws IOException {
        out.append("<moca-request autocommit=\"");
        out.append(String.valueOf(autoCommit));
        out.append("\"");
        
        if (remote) {
            out.append(" remote=\"true\"");
        }
        
        out.append(">");
        if (sessionId != null) {
            out.append("<session id=\"");
            out.append(sessionId);
            out.append("\"/>");
        }
        if (envString != null) {
            out.append("<environment>");
            
            out.append(envString);
            out.append("</environment>");
        }
        
        if (context != null) {
            out.append("<context>");
            writeArgs(context, out);
            out.append("</context>");
        }
        
        if (args != null) {
            out.append("<args>");
            writeArgs(args, out);
            out.append("</args>");
        }
        
        if (command != null) {
            out.append("<query>");
            escapeXML(command, out);
            out.append("</query>");
        }
        
        out.append("</moca-request>");
    }
    
    public static String buildXMLEnvironmentString(Map<String, String> env) {
        StringBuilder buf = new StringBuilder();
        for (Map.Entry<String, String> var : env.entrySet()) {
            String name = var.getKey();
            String value = var.getValue();
            
            try {
                buf.append("<var name=\"");
                escapeXML(name.toUpperCase(), buf);
                buf.append("\" value=\"");
                escapeXML(value, buf);
                buf.append("\"/>");
            }
            catch (IOException e) {
                // Could be thrown by escapeXML, but won't because we're writing
                // to a StringBuilder.
            }
        }
        return buf.toString();
    }
    
    private static void writeArgs(MocaArgument[] args, final Appendable out) throws IOException {
        
        for (MocaArgument arg : args) {
            String name = arg.getName();
            Object value = arg.getValue();
            MocaType type = arg.getType();
            MocaOperator oper = arg.getOper();

            out.append("<field name=\"");
            escapeXML(name, out);
            out.append("\" type=\"" + type.toString() + "\"");
            out.append(" oper=\"" + oper.toString() + "\"");
            
            if (value == null) {
                out.append(" null=\"true\"/>");
            }
            else {
                switch(type) {
                case BOOLEAN:
                case STRING:
                case INTEGER:
                case DOUBLE:
                    out.append(">");
                    escapeXML(String.valueOf(value), out);
                    break;
                case DATETIME:
                    out.append(">");
                    out.append(DateUtils.formatDate((Date)value));
                    break;
                case RESULTS:
                    out.append(">");
                    // We have to escape all the text so the tags don't get interpreted
                    // again.
                    Appendable app = new Appendable() {
                        
                        @Override
                        public Appendable append(CharSequence csq, int start, int end)
                                throws IOException {
                            escapeXML(csq.subSequence(start, end), out);
                            return this;
                        }
                        
                        @Override
                        public Appendable append(char c) throws IOException {
                            escapeXML(Character.toString(c), out);
                            return this;
                        }
                        
                        @Override
                        public Appendable append(CharSequence csq) throws IOException {
                            escapeXML(csq, out);
                            return this;
                        }
                    };
                    XMLResultsEncoder.writeResults((MocaResults)value, app);
                    break;
                case BINARY:
                default:
                    out.append(">");
                    break;
                }
                out.append("</field>");
            }
        }

    }

    private static void escapeXML(CharSequence s, Appendable out) throws IOException {
        if (s != null) {
            int length = s.length();
            for (int i = 0; i < length; i++) {
                char c = s.charAt(i);
                switch (c) {
                case '&':
                    out.append("&amp;");
                    break;
                case '<':
                    out.append("&lt;");
                    break;
                case '>':
                    out.append("&gt;");
                    break;
                case '"':
                    out.append("&quot;");
                    break;
                case '\'':
                    out.append("&apos;");
                    break;
                default:
                    out.append(c);
                    break;
                }
            }
        }
    }
}
